src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,811 @@
+            if !self.is_column_filled[self.handle_index] {
+                let handle_id = table::decode_handle(key)?;
+                // FIXME: The columns may be not in the same length if there is error.
+                // TODO: We should avoid calling `push_int` repeatly. Instead we should specialize
+                // a `&mut Vec` first.
+                columns[self.handle_index]
+                    .mut_decoded()
+                    .push_int(Some(handle_id));
+                decoded_columns += 1;
+                self.is_column_filled[self.handle_index] = true;
+            }
+            // TODO: Shall we just returns error when met
+            // `self.is_column_filled[self.handle_index] == true`?
+        }
+
+        if value.is_empty() || (value.len() == 1 && value[0] == datum::NIL_FLAG) {
+            // Do nothing
+        } else {
+            // The layout of value is: [col_id_1, value_1, col_id_2, value_2, ...]
+            // where each element is datum encoded.
+            // The column id datum must be in var i64 type.
+            let mut remaining = value;
+            while !remaining.is_empty() && decoded_columns < columns_len {
+                if remaining[0] != datum::VAR_INT_FLAG {
+                    return Err(box_err!("Expect VAR_INT flag for column id"));
+                }
+                remaining = &remaining[1..];
+                let column_id = box_try!(number::decode_var_i64(&mut remaining));
+                let (val, new_remaining) = datum::split_datum(remaining, false)?;
+                // Note: The produced columns may be not in the same length if there is error due
+                // to corrupted data. It will be handled in `ScanExecutor`.
+                let some_index = self.column_id_index.get(&column_id);
+                if let Some(index) = some_index {
+                    let index = *index;
+                    if !self.is_column_filled[index] {
+                        columns[index].push_raw(val);
+                        decoded_columns += 1;
+                        self.is_column_filled[index] = true;
+                    }
+                    // TODO: Shall we just returns error when met
+                    // `self.is_column_filled[index] == true`?
+                }
+                remaining = new_remaining;
+            }
+        }
+
+        // Some fields may be missing in the row, we push corresponding default value to make all
+        // columns in same length.
+        for i in 0..columns_len {
+            if !self.is_column_filled[i] {
+                // Missing fields must not be a primary key, so it must be
+                // `LazyBatchColumn::raw`.
+
+                let default_value = if !self.columns_default_value[i].is_empty() {
+                    self.columns_default_value[i].as_slice()
+                } else if !self.schema[i]
+                    .flag()
+                    .contains(cop_datatype::FieldTypeFlag::NOT_NULL)
```

Comment:
column with not null and doesn't have default value may need backfill zero value.
```sql
mysql root@127.0.0.1:test> create table t (a int);
Query OK, 0 rows affected
Time: 0.013s
mysql root@127.0.0.1:test> insert into t set a=1;
Query OK, 1 row affected
Time: 0.002s
mysql root@127.0.0.1:test> alter table t add column b int not null;
Query OK, 0 rows affected
Time: 0.023s
mysql root@127.0.0.1:test> select * from t
+---+---+
| a | b |
+---+---+
| 1 | 0 |
+---+---+
1 row in set
```